File: code/apps/pyobjc/Managed Software Center/Managed Software Center/MSCStatusController.py
# encoding: utf-8
#
# MSCStatusController.py
#
# Copyright 2009-2019 Greg Neagle.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


from objc import YES, IBOutlet, nil
from PyObjCTools import AppHelper
#from Foundation import *
#from AppKit import *

# pylint: disable=wildcard-import
from CocoaWrapper import *
# pylint: enable=wildcard-import

import munki
import msclog

debug = False

class MSCStatusController(NSObject):
    '''
    Handles status messages from managedsoftwareupdate
    '''

    session_started = False
    got_status_update = False
    timer = None

    _status_stopBtnDisabled = False
    _status_stopBtnHidden = False
    _status_message = u''
    _status_detail = u''
    _status_percent = -1
    _status_stopBtnState = 0

    statusWindowController = IBOutlet()

    def registerForNotifications(self):
        '''Register for notification messages'''
        notification_center = NSDistributedNotificationCenter.defaultCenter()
        notification_center.addObserver_selector_name_object_suspensionBehavior_(
            self,
            self.updateStatus_,
            'com.googlecode.munki.managedsoftwareupdate.statusUpdate',
            None,
            NSNotificationSuspensionBehaviorDeliverImmediately)
        self.receiving_notifications = True

    def unregisterForNotifications(self):
        '''Tell the DistributedNotificationCenter to stop sending us notifications'''
        NSDistributedNotificationCenter.defaultCenter().removeObserver_(self)
        # set self.receiving_notifications to False so our process monitoring
        # thread will exit
        self.receiving_notifications = False

    def startMunkiStatusSession(self):
        '''Initialize things for monitoring a managedsoftwareupdate session'''
        self.initStatusSession()
        self.session_started = True
        # start our process monitor timer so we can be notified about
        # process failure
        self.timeout_counter = 6
        self.saw_process = False
        self.timer = NSTimer.scheduledTimerWithTimeInterval_target_selector_userInfo_repeats_(
            5.0, self, self.checkProcess_, None, YES)

    def checkProcess_(self, timer):
        '''Monitors managedsoftwareupdate process for failure to start
        or unexpected exit, so we're not waiting around forever if
        managedsoftwareupdate isn't running.'''
        PYTHON_SCRIPT_NAME = u'managedsoftwareupdate'
        NEVER_STARTED = -2
        UNEXPECTEDLY_QUIT = -1

        if self.session_started:
            if self.got_status_update:
                # we got a status update since we last checked; no need to
                # check the process table
                self.timeout_counter = 6
                self.saw_process = True
                # clear the flag so we have to get another status update
                self.got_status_update = False
            elif munki.pythonScriptRunning(PYTHON_SCRIPT_NAME):
                self.timeout_counter = 6
                self.saw_process = True
            else:
                msclog.debug_log('managedsoftwareupdate not running...')
                self.timeout_counter -= 1
            if self.timeout_counter == 0:
                msclog.debug_log('Timed out waiting for managedsoftwareupdate.')
                if self.saw_process:
                    self.sessionEnded_(UNEXPECTEDLY_QUIT)
                else:
                    self.sessionEnded_(NEVER_STARTED)

    def sessionStarted(self):
        '''Accessor method'''
        return self.session_started

    def sessionEnded_(self, result):
        '''clean up after a managedsoftwareupdate session ends'''
        if self.timer:
            self.timer.invalidate()
            self.timer = None
        self.cleanUpStatusSession()
        # tell the window controller the update session is done
        self.statusWindowController.munkiStatusSessionEndedWithStatus_errorMessage_(result, "")

    def updateStatus_(self, notification):
        '''Got update status notification from managedsoftwareupdate'''
        msclog.debug_log('Got munkistatus update notification')
        self.got_status_update = True
        info = notification.userInfo()
        msclog.debug_log('%s' % info)
        # explicitly get keys from info object; PyObjC in Mountain Lion
        # seems to need this
        info_keys = info.keys()
        if 'message' in info_keys:
            self.setMessage_(info['message'])
        if 'detail' in info_keys:
            self.setDetail_(info['detail'])
        if 'percent' in info_keys:
            self.setPercentageDone_(info['percent'])
        if 'stop_button_visible' in info_keys:
            if info['stop_button_visible']:
                self.showStopButton()
            else:
                self.hideStopButton()
        if 'stop_button_enabled' in info_keys:
            if info['stop_button_enabled']:
                self.enableStopButton()
            else:
                self.disableStopButton()

        command = info.get('command')
        if not self.session_started and command not in ['showRestartAlert', 'quit']:
            # we got a status message but we didn't start the session
            # so switch to the right mode
            self.startMunkiStatusSession()
        if command:
            msclog.debug_log('Received command: %s' % command)
        if command == 'activate':
            pass

        elif command == 'showRestartAlert':
            if self.session_started:
                self.sessionEnded_(0)
            self.doRestartAlert()
        elif command == 'quit':
            self.sessionEnded_(0)

##### required status methods #####

    def initStatusSession(self):
        '''Initialize the main window for update status'''
        self.statusWindowController._update_in_progress = True
        if self.statusWindowController.currentPageIsUpdatesPage():
            self.statusWindowController.webView.reload_(self)
            self.statusWindowController.displayUpdateCount()

    def cleanUpStatusSession(self):
        '''Clean up after status session ends'''
        self.session_started = False
        # reset all our status variables
        self.statusWindowController._update_in_progress = False
        self._status_stopBtnDisabled = False
        self._status_stopBtnHidden = False
        self._status_stopBtnState = 0
        self._status_message = u''
        self._status_detail = u''
        self._status_percent = -1

    def setPercentageDone_(self, percent):
        '''Display percentage done'''
        try:
            if float(percent) > 100.0:
                percent = 100
        except ValueError:
            percent = 0
        self._status_percent = percent
        document = self.statusWindowController.webView.mainFrameDocument()
        if document:
            spinner = document.getElementById_('updates-progress-spinner')
            if spinner: # we are displaying the updates status page
                progress = document.getElementById_('progress-bar')
                if progress:
                    if float(percent) < 0:
                        # indeterminate
                        progress.setClassName_('indeterminate')
                        progress.removeAttribute_('style')
                    else:
                        progress.setClassName_('')
                        progress.setAttribute__(
                            'style', 'width: %s%%' % percent)

    def doRestartAlert(self):
        '''Display a restart alert -- some item just installed or removed
        requires a restart'''
        msclog.log("MSC", "restart_required")
        self._status_restartAlertDismissed = 0
        alert = NSAlert.alertWithMessageText_defaultButton_alternateButton_otherButton_informativeTextWithFormat_(
            NSLocalizedString(u"Restart Required", u"Restart Required title"),
            NSLocalizedString(u"Restart", u"Restart button title"),
            nil,
            nil,
            u"%@", NSLocalizedString(
                u"Software installed or removed requires a restart. You will "
                "have a chance to save open documents.",
                u"Restart Required alert detail"))
        alert.beginSheetModalForWindow_modalDelegate_didEndSelector_contextInfo_(
            self.statusWindowController.window(),
            self, self.restartAlertDidEnd_returnCode_contextInfo_, nil)

    @AppHelper.endSheetMethod
    def restartAlertDidEnd_returnCode_contextInfo_(
            self, alert, returncode, contextinfo):
        '''Called when restartAlert ends'''
        msclog.log("MSC", "restart_confirmed")
        self._status_restartAlertDismissed = 1
        munki.restartNow()

    def setMessage_(self, messageText):
        '''Display main status message'''
        messageText = NSBundle.mainBundle().localizedStringForKey_value_table_(
            messageText, messageText, None)
        self._status_message = messageText
        document = self.statusWindowController.webView.mainFrameDocument()
        if document:
            spinner = document.getElementById_('updates-progress-spinner')
            if spinner: # we are displaying the updates status page
                textElement = document.getElementById_('primary-status-text')
                if textElement:
                    if messageText:
                        textElement.setInnerText_(messageText)
                    else:
                        textElement.setInnerHTML_('&nbsp;')

    def setDetail_(self, detailText):
        '''Display status detail'''
        detailText = NSBundle.mainBundle().localizedStringForKey_value_table_(
            detailText, detailText, None)
        self._status_detail = detailText
        document = self.statusWindowController.webView.mainFrameDocument()
        if document:
            spinner = document.getElementById_('updates-progress-spinner')
            if spinner: # we are displaying the updates status page
                textElement = document.getElementById_('secondary-status-text')
                if textElement:
                    if detailText:
                        textElement.setInnerText_(detailText)
                    else:
                        textElement.setInnerHTML_('&nbsp;')

    def getStopBtnState(self):
        '''Get the state (pressed or not) of the stop button'''
        return self._status_stopBtnState

    def hideStopButton(self):
        '''Hide the stop button'''
        if self._status_stopBtnState:
            return
        self._status_stopBtnHidden = True
        document = self.statusWindowController.webView.mainFrameDocument()
        if document:
            spinner = document.getElementById_('updates-progress-spinner')
            if spinner: # we are displaying the updates status page
                install_btn = document.getElementById_(
                    'install-all-button-text')
                if install_btn:
                    btn_classes = install_btn.className().split(' ')
                    if not 'hidden' in btn_classes:
                        btn_classes.append('hidden')
                        install_btn.setClassName_(' '.join(btn_classes))

    def showStopButton(self):
        '''Show the stop button'''
        if self._status_stopBtnState:
            return
        self._status_stopBtnHidden = False
        document = self.statusWindowController.webView.mainFrameDocument()
        if document:
            spinner = document.getElementById_('updates-progress-spinner')
            if spinner: # we are displaying the updates status page
                install_btn = document.getElementById_(
                    'install-all-button-text')
                if install_btn:
                    btn_classes = install_btn.className().split(' ')
                    if 'hidden' in btn_classes:
                        btn_classes.remove('hidden')
                        install_btn.setClassName_(' '.join(btn_classes))

    def enableStopButton(self):
        '''Enable the stop button'''
        if self._status_stopBtnState:
            return
        self._status_stopBtnDisabled = False
        document = self.statusWindowController.webView.mainFrameDocument()
        if document:
            spinner = document.getElementById_('updates-progress-spinner')
            if spinner: # we are displaying the updates status page
                install_btn = document.getElementById_(
                    'install-all-button-text')
                if install_btn:
                    btn_classes = install_btn.className().split(' ')
                    if 'disabled' in btn_classes:
                        btn_classes.remove('disabled')
                        install_btn.setClassName_(' '.join(btn_classes))

    def disableStopButton(self):
        '''Disable the stop button'''
        if self._status_stopBtnState:
            return
        self._status_stopBtnDisabled = True
        document = self.statusWindowController.webView.mainFrameDocument()
        if document:
            spinner = document.getElementById_('updates-progress-spinner')
            if spinner: # we are displaying the updates status page
                install_btn = document.getElementById_(
                    'install-all-button-text')
                if install_btn:
                    btn_classes = install_btn.className().split(' ')
                    if not 'disabled' in btn_classes:
                        btn_classes.append('disabled')
                        install_btn.setClassName_(' '.join(btn_classes))

    def getRestartAlertDismissed(self):
        '''Was the restart alert dismissed?'''
        return self._status_restartAlertDismissed


def more_localized_strings():
    '''Some strings that are sent to us from managedsoftwareupdate. By putting
    them here, the localize.py script will add them to the
    en.lproj/Localizable.strings file so localizers will be able to discover
    them'''
    _ = NSLocalizedString(u"Starting...", "managedsoftwareupdate message")
    _ = NSLocalizedString(u"Finishing...", "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Performing preflight tasks...", "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Performing postflight tasks...", "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Checking for available updates...", "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Checking for additional changes...", "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Software installed or removed requires a restart.",
        "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Waiting for network...", "managedsoftwareupdate message")
    _ = NSLocalizedString(u"Done.", "managedsoftwareupdate message")

    _ = NSLocalizedString(
        u"Retrieving list of software for this machine...",
        "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Verifying package integrity...", "managedsoftwareupdate message")

    _ = NSLocalizedString(u"The software was successfully installed.",
                          "managedsoftwareupdate message")

    _ = NSLocalizedString(u"Gathering information on installed packages",
                          "managedsoftwareupdate message")
    _ = NSLocalizedString(u"Determining which filesystem items to remove",
                          "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Removing receipt info", "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Nothing to remove.", "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Package removal complete.", "managedsoftwareupdate message")

    _ = NSLocalizedString(u"Checking for available Apple Software Updates...",
                          "managedsoftwareupdate message")
    _ = NSLocalizedString(u"Checking Apple Software Update catalog...",
                          "managedsoftwareupdate message")
    _ = NSLocalizedString(u"Downloading available Apple Software Updates...",
                          "managedsoftwareupdate message")
    _ = NSLocalizedString(u"Installing available Apple Software Updates...",
                          "managedsoftwareupdate message")

    _ = NSLocalizedString(
        u"Running Adobe Setup", "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Running Adobe Uninstall", "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Starting Adobe installer...", "managedsoftwareupdate message")
    _ = NSLocalizedString(
        u"Running Adobe Patch Installer", "managedsoftwareupdate message")

File: code/apps/Managed Software Center/Managed Software Center/Controllers/MSCStatusController.swift
//
//  MSCStatusController.swift
//  Managed Software Center
//
//  Created by Greg Neagle on 7/11/18.
//  Copyright © 2018-2026 The Munki Project. All rights reserved.
//

import Cocoa
import WebKit

class MSCStatusController: NSObject {
    
    @IBOutlet weak var statusWindowController: MainWindowController!
    
    // Handles status messages from managedsoftwareupdate
    var session_started = false
    var got_status_update = false
    var timer: Timer? = nil
    
    var _status_restartAlertDismissed = false
    var _status_stopBtnDisabled = false
    var _status_stopBtnHidden = false
    var _status_stopBtnState = 0
    var _status_message = ""
    var _status_detail = ""
    var _status_percent = -1.0
    var _status_stopBtnPressed = false
    var receiving_notifications = false
    var timeout_counter = 0
    var saw_process = false
    
    func registerForNotifications() {
        // Register for notification messages
        let center = DistributedNotificationCenter.default()
        center.addObserver(self,
                           selector: #selector(self.updateStatus),
                           name: NSNotification.Name(
                                rawValue: "com.googlecode.munki.managedsoftwareupdate.statusUpdate"),
                           object: nil,
                           suspensionBehavior: .deliverImmediately)
        receiving_notifications = true
    }
    
    func unregisterForNotifications() {
        // Tell the DistributedNotificationCenter to stop sending us notifications
        DistributedNotificationCenter.default().removeObserver(self)
        // set self.receiving_notifications to False so our process monitoring
        // thread will exit
        receiving_notifications = false
    }
    
    func startMunkiStatusSession() {
        // Initialize things for monitoring a managedsoftwareupdate session
        initStatusSession()
        session_started = true
        // start our process monitor timer so we can be notified about
        // process failure
        timeout_counter = 6
        saw_process = false
        timer = Timer.scheduledTimer(timeInterval: 5.0,
                                     target: self,
                                     selector: #selector(self.checkProcess),
                                     userInfo: nil,
                                     repeats: true)
    }
    
    @objc func checkProcess() {
        // Monitors managedsoftwareupdate process for failure to start
        // or unexpected exit, so we're not waiting around forever if
        // managedsoftwareupdate isn't running.
        let NEVER_STARTED = -2
        let UNEXPECTEDLY_QUIT = -1
        if !session_started {
            return
        }
        if got_status_update {
            // we got a status update since we last checked; no need to
            // check the process table
            timeout_counter = 6
            saw_process = true
            // clear the flag so we have to get another status update
            got_status_update = false
        } else if managedsoftwareupdateInstanceRunning() {
            timeout_counter = 6
            saw_process = true
        } else {
            msc_debug_log("managedsoftwareupdate not running...")
            timeout_counter -= 1
        }
        if timeout_counter == 0 {
            msc_debug_log("Timed out waiting for managedsoftwareupdate.")
            if saw_process {
                sessionEnded(UNEXPECTEDLY_QUIT)
            } else {
                sessionEnded(NEVER_STARTED)
            }
        }
    }
    
    func sessionStarted() -> Bool {
        // Accessor method
        return session_started
    }
    
    func sessionEnded(_ result: Int) {
        // clean up after a managedsoftwareupdate session ends
        if let uTimer = timer {
            uTimer.invalidate()
            timer = nil
        }
        cleanUpStatusSession()
        // tell the window controller the update session is done
        statusWindowController.munkiStatusSessionEnded(withStatus: result, errorMessage: "")
    }
    
    @objc func updateStatus(_ notification: NSNotification) {
        // Got update status notification from managedsoftwareupdate
        msc_debug_log("Got munkistatus update notification")
        got_status_update = true
        guard let info = notification.userInfo as? [String: Any] else {
            msc_debug_log("No userInfo in notification")
            return
        }
        msc_debug_log("\(info)")
        if let message = info["message"] as? String {
            setMessage(message)
        }
        if let detail = info["detail"] as? String {
            setDetail(detail)
        }
        if let percent = info["percent"] {
            setPercentageDone(percent)
        }
        if let stop_button_visible = info["stop_button_visible"] as? Bool {
            if stop_button_visible {
                showStopButton()
            } else {
                hideStopButton()
            }
        }
        if let stop_button_enabled = info["stop_button_enabled"] as? Bool {
            if stop_button_enabled {
                enableStopButton()
            } else {
                disableStopButton()
            }
        }
        let command = info["command"] as? String ?? ""
        if !session_started && !["showRestartAlert", "quit"].contains(command) {
            // we got a status message but we didn't start the session
            // so switch to the right mode
            startMunkiStatusSession()
        }
        if !command.isEmpty {
            msc_debug_log("Received command: \(command)")
        }
        if command == "activate" {
            // do nothing
        } else if command == "showRestartAlert" {
            if session_started {
                sessionEnded(0)
            }
            doRestartAlert()
        } else if command == "quit" {
            sessionEnded(0)
        }
    }
    
    // required status methods

    func initStatusSession() {
        // Initialize the main window for update status
        statusWindowController._update_in_progress = true
        statusWindowController.displayUpdatesProgressSpinner(true)
        if statusWindowController.currentPageIsUpdatesPage() {
            //statusWindowController.makeUsUnobnoxious()
            statusWindowController.load_page("updates.html")
            statusWindowController.displayUpdateCount()
        }
    }
    
    func cleanUpStatusSession() {
        // Clean up after status session ends
        session_started = false
        // stop the progress spinner in the sidebar
        statusWindowController.displayUpdatesProgressSpinner(false)
        // reset all our status variables
        statusWindowController._update_in_progress = false
        _status_stopBtnDisabled = false
        _status_stopBtnHidden = false
        _status_stopBtnState = 0
        _status_message = ""
        _status_detail = ""
        _status_percent = -1.0
    }
    
    func doRestartAlert() {
        // Display a restart alert -- some item just installed or removed
        // requires a restart
        msc_log("MSC", "restart_required")
        _status_restartAlertDismissed = false
        let alert = NSAlert()
        alert.messageText = NSLocalizedString("Restart Required", comment: "Restart Required title")
        alert.informativeText = NSLocalizedString(
            "Software installed or removed requires a restart. You will have a chance to save open documents.",
            comment:"Restart Required alert detail")
        alert.addButton(withTitle: NSLocalizedString("Restart", comment:"Restart button title"))
        alert.beginSheetModal(for: statusWindowController.window!) { (_) in
            msc_log("MSC", "restart_confirmed")
            self._status_restartAlertDismissed = true
            restartNow()
        }
    }

    func setPercentageDone(_ percentish: Any?) {
        // Display percentage done
        var percent = -1.0
        // TO-DO: there's got to be a more elegant way of doing this
        if percentish != nil {
            if percentish! is String {
                percent = Double(percentish! as! String) ?? -1.0
            } else if percentish! is Double {
                percent = percentish! as! Double
            }
        }
        if percent > 100.0 {
            percent = 100.0
        }
        _status_percent = percent
        // tell JavaScript to update the percentage done
        msc_debug_log("Calling JavaScript updateProgress")
        statusWindowController.webView.evaluateJavaScript("updateProgress(\(percent))")
    }

    func setMessage(_ text: String) {
        // Display main status message
        var messageText = Bundle.main.localizedString(forKey: text, value: text, table: nil)
        _status_message = messageText
        if messageText.isEmpty {
            messageText = "&nbsp;"
        }
        statusWindowController.setInnerHTML(messageText, elementID: "primary-status-text")
    }
    
    func setDetail(_ text: String) {
        // Display status detail
        var detailText = Bundle.main.localizedString(forKey: text, value: text, table: nil)
        _status_detail = detailText
        if detailText.isEmpty {
            detailText = "&nbsp;"
        }
        statusWindowController.setInnerHTML(detailText, elementID: "secondary-status-text")
    }

    func getStopButtonState() -> Bool {
        // Get the state (pressed or not) of the stop button
        // Returns true if pressed; false if not
        return _status_stopBtnPressed
    }
    
    func hideStopButton() {
        // Hide the stop button
        if _status_stopBtnPressed {
            // if the button is pressed, don't hide it
            return
        }
        _status_stopBtnHidden = true
        statusWindowController.webView.evaluateJavaScript("hideStopButton()")
    }
    
    func showStopButton() {
        // Show the stop button
        if _status_stopBtnPressed {
            // if the button is pressed, just exit
            return
        }
        _status_stopBtnHidden = false
        statusWindowController.webView.evaluateJavaScript("showStopButton()")
    }
    
    func disableStopButton() {
        // Disable the stop button
        if _status_stopBtnPressed {
            // if the button is pressed, just exit
            return
        }
        _status_stopBtnHidden = true
        statusWindowController.webView.evaluateJavaScript("disableStopButton()")
    }
    
    func enableStopButton() {
        // Enable the stop button
        if _status_stopBtnPressed {
            // if the button is pressed, just exit
            return
        }
        _status_stopBtnDisabled = false
        statusWindowController.webView.evaluateJavaScript("enableStopButton()")
    }
    
    func getRestartAlertDismissed() -> Bool {
        // Was the restart alert dismissed?'
        return _status_restartAlertDismissed
    }
    
    func morelocalizedstrings() {
        // Some strings that are sent to us from managedsoftwareupdate. By putting
        // them here, genstrings can add them to the Localizable.strings file
        // so localizers will be able to discover them
        
        var  _ = "" // we don't actually use these values at all
        
        // Munki messages
        _ = NSLocalizedString(
            "Starting...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Finishing...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Performing preflight tasks...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Performing postflight tasks...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Checking for available updates...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Checking for additional changes...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Software installed or removed requires a restart.",
            comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Waiting for network...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString("Done.", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Retrieving list of software for this machine...",
            comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Verifying package integrity...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "The software was successfully installed.",
            comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Gathering information on installed packages",
            comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Determining which filesystem items to remove",
            comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Removing receipt info", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Nothing to remove.", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Package removal complete.", comment: "managedsoftwareupdate message")
        
        // apple update messages
        _ = NSLocalizedString(
            "Checking for available Apple Software Updates...",
            comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Checking Apple Software Update catalog...",
            comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Downloading available Apple Software Updates...",
            comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Installing available Apple Software Updates...",
            comment: "managedsoftwareupdate message")
        
        // Adobe install/uninstall messages
        _ = NSLocalizedString(
            "Running Adobe Setup", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Running Adobe Uninstall", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Starting Adobe installer...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Running Adobe Patch Installer", comment: "managedsoftwareupdate message")
        
        // macOS install/upgrade messages
        _ = NSLocalizedString(
            "Starting macOS upgrade...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "Preparing to run macOS Installer...", comment: "managedsoftwareupdate message")
        _ = NSLocalizedString(
            "System will restart and begin upgrade of macOS.",
            comment: "managedsoftwareupdate message")
    }

}
